pydrofoil/real.py

```diff
@@ -0,0 +1,264 @@
+        return Real(num_new, den_new)
+    
+    def mul(self, other):
+        num_new = self.num.mul(other.num)
+        den_new = self.den.mul(other.den)
+        return Real(num_new, den_new)
+    
+    def div(self, other):
+        if self.num.int_eq(0) and not other.num.int_eq(0):
+            num_new = rbigint.fromint(0)
+            den_new = rbigint.fromint(1)
+        elif other.num.int_eq(0):
+            assert False, "ZerodivideError: denominator cannot be 0"
+        else:
+            num_new = self.num.mul(other.den)
+            den_new = self.den.mul(other.num)
+        if den_new.int_lt(0):
+            num_new = num_new.neg()
+            den_new = den_new.neg()
+        return Real(num_new, den_new)
+
+    def pow(self, n):
+        if isinstance(n, int):
+            if n == 0:
+                return Real(rbigint.fromint(1), rbigint.fromint(1))
+            elif n < MININT or n > MAXINT:
+                assert False, "exponent is out of range of INT"
+            else:
+                n = rbigint.fromint(n)
+                num_new = self.num.pow(n)
+                den_new = self.den.pow(n)
+                return Real(num_new, den_new)
+        elif isinstance(n, rbigint):
+            if n.int_eq(0):
+                return Real(rbigint.fromint(1), rbigint.fromint(1))
+            else:
+                n = rbigint.fromint(n)
+                num_new = self.num.pow(n)
+                den_new = self.den.pow(n)
+                return Real(num_new, den_new)
+        elif isinstance(n, Real):
+            if n.num.int_eq(0):
+                return Real(rbigint.fromint(1), rbigint.fromint(1))
+            else:
+                if n.den.int_eq(1):
+                    num_new = self.num.pow(n.num)
+                    den_new = self.den.pow(n.num)
+                    return Real(num_new, den_new)
+                else:
+                    assert False, "exponent cannot be fraction"
+        else:
+            raise TypeError("exponent doesn't support this type")
+        
+        
+            
+    
+    def neg(self):
+        return Real(self.num.neg(), self.den)
+    
+    def abs(self):
+        return Real(self.num.abs(), self.den)
+
+    def ceil(self):
+        diff = self.num.divmod(self.den)
+        num_new = self.num if diff[1].int_eq(0) else self.num.add(self.den.sub(diff[1]))
+        return Real(num_new, self.den)
+
+    def floor(self):
+        diff = self.num.divmod(self.den)
+        num_new = self.num if diff[1].int_eq(0) else self.num.sub(diff[1])
+        return Real(num_new, self.den)
+    
+    def eq(self, other):
+        return self.num.eq(other.num) and self.den.eq(other.den)
+    
+    def lt(self, other):
+        if (self.den.int_gt(0) and other.den.int_gt(0)) or (not self.den.int_gt(0) and not other.den.int_gt(0)):
+            return self.num.mul(other.den).lt(self.den.mul(other.num))
+        elif self.den.int_gt(0):
+            return False
+        else:
```

Comment:
I don't understand the need for several cases here. Shouldn't it just always be this:

```python
return self.num.mul(other.den).lt(self.den.mul(other.num))
```

Reply:
oh yes, sorry, that's my fault, cause before I put the sign to the denominator, so I have to consider about different cases, but now since the sign is on numerator, it's always this